File: sdks/go/dcp/v2/pq_checkpoint.go
// Package v2 — DCP v2.0 Lazy PQ Checkpoint (Go port).

package v2

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
)

// AuditEventsMerkleRoot computes a SHA-256 Merkle root over canonicalised event payloads.
// Returns 64 lowercase hex chars (no "sha256:" prefix).
func AuditEventsMerkleRoot(events []map[string]interface{}) (string, error) {
	if len(events) == 0 {
		return "", errors.New("cannot compute Merkle root of empty event list")
	}
	leaves := make([]string, 0, len(events))
	for _, e := range events {
		c, err := CanonicalizeV2(e)
		if err != nil {
			return "", fmt.Errorf("canonicalize: %w", err)
		}
		sum := sha256.Sum256([]byte(c))
		leaves = append(leaves, hex.EncodeToString(sum[:]))
	}
	for len(leaves) > 1 {
		if len(leaves)%2 == 1 {
			leaves = append(leaves, leaves[len(leaves)-1])
		}
		next := make([]string, 0, len(leaves)/2)
		for i := 0; i < len(leaves); i += 2 {
			l, err := hex.DecodeString(leaves[i])
			if err != nil {
				return "", err
			}
			r, err := hex.DecodeString(leaves[i+1])
			if err != nil {
				return "", err
			}
			combined := append(l, r...)
			sum := sha256.Sum256(combined)
			next = append(next, hex.EncodeToString(sum[:]))
		}
		leaves = next
	}
	return leaves[0], nil
}

// CreatePQCheckpoint signs a Merkle-root checkpoint over a batch of events.
func CreatePQCheckpoint(
	registry *AlgorithmRegistry,
	classicalKey CompositeKeyInfo,
	pqKey CompositeKeyInfo,
	events []map[string]interface{},
	sessionNonce string,
) (map[string]interface{}, error) {
	if len(events) == 0 {
		return nil, errors.New("cannot create checkpoint for empty event list")
	}
	root, err := AuditEventsMerkleRoot(events)
	if err != nil {
		return nil, err
	}
	fromID, _ := events[0]["audit_id"].(string)
	toID, _ := events[len(events)-1]["audit_id"].(string)

	payload := map[string]interface{}{
		"checkpoint_id": "ckpt-" + newUUIDv4(),
		"session_nonce": sessionNonce,
		"event_range": map[string]interface{}{
			"from_audit_id": fromID,
			"to_audit_id":   toID,
			"count":         len(events),
		},
		"merkle_root": "sha256:" + root,
	}
	canonical, err := CanonicalizeV2(payload)
	if err != nil {
		return nil, fmt.Errorf("canonicalize: %w", err)
	}
	sig, err := CompositeSign(registry, CtxAuditEvent, []byte(canonical), classicalKey, pqKey)
	if err != nil {
		return nil, err
	}
	payload["composite_sig"] = sig
	return payload, nil
}

func newUUIDv4() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf(
		"%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
		b[0], b[1], b[2], b[3], b[4], b[5],
		b[6], b[7], b[8], b[9],
		b[10], b[11], b[12], b[13], b[14], b[15],
	)
}

// PQCheckpointManager batches audit events and flushes a PQ checkpoint at interval.
type PQCheckpointManager struct {
	registry     *AlgorithmRegistry
	classicalKey CompositeKeyInfo
	pqKey        CompositeKeyInfo
	sessionNonce string

	interval    int
	tier        string
	pending     []map[string]interface{}
	checkpoints []map[string]interface{}
}

// NewPQCheckpointManager constructs a manager with an explicit interval.
func NewPQCheckpointManager(
	interval int,
	registry *AlgorithmRegistry,
	sessionNonce string,
	classicalKey, pqKey CompositeKeyInfo,
) (*PQCheckpointManager, error) {
	if interval < 1 {
		return nil, errors.New("checkpoint interval must be >= 1")
	}
	return &PQCheckpointManager{
		registry:     registry,
		classicalKey: classicalKey,
		pqKey:        pqKey,
		sessionNonce: sessionNonce,
		interval:     interval,
	}, nil
}

// NewPQCheckpointManagerWithTier constructs a manager driven by a security tier.
func NewPQCheckpointManagerWithTier(
	tier string,
	registry *AlgorithmRegistry,
	sessionNonce string,
	classicalKey, pqKey CompositeKeyInfo,
) *PQCheckpointManager {
	return &PQCheckpointManager{
		registry:     registry,
		classicalKey: classicalKey,
		pqKey:        pqKey,
		sessionNonce: sessionNonce,
		interval:     TierToCheckpointInterval(tier),
		tier:         tier,
	}
}

// Interval returns the current checkpoint interval.
func (m *PQCheckpointManager) Interval() int { return m.interval }

// Tier returns the current tier ("" if explicit interval was used).
func (m *PQCheckpointManager) Tier() string { return m.tier }

// SetTier updates the tier and recomputes the interval; does not flush pending.
func (m *PQCheckpointManager) SetTier(tier string) {
	m.tier = tier
	m.interval = TierToCheckpointInterval(tier)
}

// PendingCount returns the number of un-checkpointed events.
func (m *PQCheckpointManager) PendingCount() int { return len(m.pending) }

// Checkpoints returns a snapshot of emitted checkpoints.
func (m *PQCheckpointManager) Checkpoints() []map[string]interface{} {
	out := make([]map[string]interface{}, len(m.checkpoints))
	copy(out, m.checkpoints)
	return out
}

// RecordEvent appends an event; flushes if pending reaches interval.
func (m *PQCheckpointManager) RecordEvent(event map[string]interface{}) (map[string]interface{}, error) {
	m.pending = append(m.pending, event)
	if len(m.pending) >= m.interval {
		return m.Flush()
	}
	return nil, nil
}

// Flush forces a checkpoint over the current pending events (nil if none).
func (m *PQCheckpointManager) Flush() (map[string]interface{}, error) {
	if len(m.pending) == 0 {
		return nil, nil
	}
	ckpt, err := CreatePQCheckpoint(m.registry, m.classicalKey, m.pqKey, m.pending, m.sessionNonce)
	if err != nil {
		return nil, err
	}
	m.checkpoints = append(m.checkpoints, ckpt)
	m.pending = m.pending[:0]
	return ckpt, nil
}

File: sdks/rust/src/v2/pq_checkpoint.rs
//! DCP v2.0 Lazy PQ Checkpoint — Rust port.

use serde_json::{json, Value};
use sha2::{Digest, Sha256};

use crate::v2::canonicalize::canonicalize_v2;
use crate::v2::composite_ops::{composite_sign, CompositeKeyInfo};
use crate::v2::crypto_provider::{CryptoError, CryptoProvider};
use crate::v2::domain_separation::CTX_AUDIT_EVENT;
use crate::v2::security_tier::{tier_to_checkpoint_interval, SecurityTier};

/// Compute a SHA-256 Merkle root over canonicalised audit-event payloads.
/// Returns a 64-char lowercase hex string (no `sha256:` prefix).
pub fn audit_events_merkle_root(events: &[Value]) -> Result<String, String> {
    if events.is_empty() {
        return Err("Cannot compute Merkle root of empty event list".into());
    }
    let mut leaves: Vec<String> = events
        .iter()
        .map(|e| {
            let canonical = canonicalize_v2(e)?;
            let mut hasher = Sha256::new();
            hasher.update(canonical.as_bytes());
            Ok(hex::encode(hasher.finalize()))
        })
        .collect::<Result<_, String>>()?;

    while leaves.len() > 1 {
        if leaves.len() % 2 == 1 {
            leaves.push(leaves.last().unwrap().clone());
        }
        let mut next = Vec::with_capacity(leaves.len() / 2);
        let mut i = 0;
        while i < leaves.len() {
            let l = hex::decode(&leaves[i]).map_err(|e| e.to_string())?;
            let r = hex::decode(&leaves[i + 1]).map_err(|e| e.to_string())?;
            let mut combined = Vec::with_capacity(l.len() + r.len());
            combined.extend_from_slice(&l);
            combined.extend_from_slice(&r);
            let mut hasher = Sha256::new();
            hasher.update(&combined);
            next.push(hex::encode(hasher.finalize()));
            i += 2;
        }
        leaves = next;
    }
    Ok(leaves.remove(0))
}

/// Build + sign a PQ checkpoint for a batch of audit events.
pub fn create_pq_checkpoint(
    classical_provider: &dyn CryptoProvider,
    pq_provider: &dyn CryptoProvider,
    classical_key: &CompositeKeyInfo,
    pq_key: &CompositeKeyInfo,
    events: &[Value],
    session_nonce: &str,
) -> Result<Value, CryptoError> {
    if events.is_empty() {
        return Err(CryptoError::SignatureError(
            "Cannot create checkpoint for empty event list".into(),
        ));
    }
    let merkle = audit_events_merkle_root(events).map_err(CryptoError::SignatureError)?;
    let checkpoint_id = format!("ckpt-{}", generate_uuidv4());

    let from_id = events
        .first()
        .and_then(|e| e.get("audit_id"))
        .and_then(Value::as_str)
        .unwrap_or("");
    let to_id = events
        .last()
        .and_then(|e| e.get("audit_id"))
        .and_then(Value::as_str)
        .unwrap_or("");

    let payload = json!({
        "checkpoint_id": checkpoint_id,
        "session_nonce": session_nonce,
        "event_range": {
            "from_audit_id": from_id,
            "to_audit_id": to_id,
            "count": events.len(),
        },
        "merkle_root": format!("sha256:{}", merkle),
    });
    let canonical = canonicalize_v2(&payload).map_err(CryptoError::SignatureError)?;
    let composite = composite_sign(
        classical_provider,
        pq_provider,
        CTX_AUDIT_EVENT,
        canonical.as_bytes(),
        classical_key,
        pq_key,
    )?;
    let mut out = payload;
    if let Some(obj) = out.as_object_mut() {
        obj.insert(
            "composite_sig".into(),
            serde_json::to_value(&composite)
                .map_err(|e| CryptoError::SignatureError(e.to_string()))?,
        );
    }
    Ok(out)
}

fn generate_uuidv4() -> String {
    // Minimal UUIDv4 without a dep: 16 random bytes, set variant + version.
    use rand::RngCore;
    let mut bytes = [0u8; 16];
    rand::thread_rng().fill_bytes(&mut bytes);
    bytes[6] = (bytes[6] & 0x0f) | 0x40; // version 4
    bytes[8] = (bytes[8] & 0x3f) | 0x80; // variant RFC 4122
    format!(
        "{:02x}{:02x}{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
        bytes[0], bytes[1], bytes[2], bytes[3],
        bytes[4], bytes[5],
        bytes[6], bytes[7],
        bytes[8], bytes[9],
        bytes[10], bytes[11], bytes[12], bytes[13], bytes[14], bytes[15],
    )
}

/// Stateful manager that batches audit events and emits a checkpoint when the
/// pending count reaches the interval.
pub struct PQCheckpointManager {
    interval: u32,
    tier: Option<SecurityTier>,
    session_nonce: String,
    pending: Vec<Value>,
    checkpoints: Vec<Value>,
}

impl PQCheckpointManager {
    pub fn new(interval: u32, session_nonce: impl Into<String>) -> Result<Self, String> {
        if interval < 1 {
            return Err("Checkpoint interval must be >= 1".into());
        }
        Ok(Self {
            interval,
            tier: None,
            session_nonce: session_nonce.into(),
            pending: Vec::new(),
            checkpoints: Vec::new(),
        })
    }

    pub fn with_tier(tier: SecurityTier, session_nonce: impl Into<String>) -> Self {
        Self {
            interval: tier_to_checkpoint_interval(tier),
            tier: Some(tier),
            session_nonce: session_nonce.into(),
            pending: Vec::new(),
            checkpoints: Vec::new(),
        }
    }

    pub fn interval(&self) -> u32 {
        self.interval
    }
    pub fn tier(&self) -> Option<SecurityTier> {
        self.tier
    }
    pub fn set_tier(&mut self, tier: SecurityTier) {
        self.tier = Some(tier);
        self.interval = tier_to_checkpoint_interval(tier);
    }
    pub fn pending_count(&self) -> usize {
        self.pending.len()
    }
    pub fn checkpoints(&self) -> &[Value] {
        &self.checkpoints
    }

    /// Record an event and emit a checkpoint if the interval is hit.
    pub fn record_event(
        &mut self,
        classical_provider: &dyn CryptoProvider,
        pq_provider: &dyn CryptoProvider,
        classical_key: &CompositeKeyInfo,
        pq_key: &CompositeKeyInfo,
        event: Value,
    ) -> Result<Option<Value>, CryptoError> {
        self.pending.push(event);
        if self.pending.len() as u32 >= self.interval {
            return self.flush(classical_provider, pq_provider, classical_key, pq_key);
        }
        Ok(None)
    }

    /// Force a checkpoint over all pending events.
    pub fn flush(
        &mut self,
        classical_provider: &dyn CryptoProvider,
        pq_provider: &dyn CryptoProvider,
        classical_key: &CompositeKeyInfo,
        pq_key: &CompositeKeyInfo,
    ) -> Result<Option<Value>, CryptoError> {
        if self.pending.is_empty() {
            return Ok(None);
        }
        let checkpoint = create_pq_checkpoint(
            classical_provider,
            pq_provider,
            classical_key,
            pq_key,
            &self.pending,
            &self.session_nonce,
        )?;
        self.checkpoints.push(checkpoint.clone());
        self.pending.clear();
        Ok(Some(checkpoint))
    }
}
